### PR TITLE
Traj dpo

### PR DESCRIPTION
**Pull request by dvattk**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -992,7 +995,13 @@ def train(self) -> None:
                 # Compute loss
                 # Loss is normalized by default so we multiply by the number of tokens
                 # This way we can normalize by the total number of tokens if we're accumulating gradients
-                current_loss = self._loss_fn(logits, labels) * current_num_tokens
+                if self.method=="reinforce":
```
Comment (optimass) [inline recipes/full_finetune_distributed.py]: do it for single_device as well

File: torchtune/models/llama3/_tokenizer.py [hunk 2]
```diff
@@ -74,7 +75,7 @@ def __init__(
         path: str,
         special_tokens: Optional[Dict[str, int]] = None,
         max_seq_len: Optional[int] = None,
-        prompt_template: Optional[PromptTemplate] = None,
+        prompt_template: Optional[PromptTemplate] = Llama3ChatTemplate(),
```
Comment (optimass) [inline torchtune/models/llama3/_tokenizer.py:78]: create a model_to_chat_template dict, and pass the prompt_template in tune_run
